### PR TITLE
--wait, wait for more components 

### PR DESCRIPTION
**Pull request by medyagh**
redoing this PR https://github.com/kubernetes/minikube/pull/7209
adding a new component (default service account) to wait for. 
based suggestions in this comment https://github.com/kubernetes/minikube/pull/7209#issuecomment-605150448
```
minikube start --wait=false: wait for nothing, same as today
minikube start --wait=responses -- default: no behavioral change from today
minikube start --wait=all -- wait for everything

minikube start --wait=true -- alias for "all" for backwards compatibility
```

closes https://github.com/kubernetes/minikube/issues/7011

## Before this PR:
```
      --wait=true: Block until the apiserver is servicing API requests
```
## After this PR:
```
      --wait=[apiserver,system_pods]: comma separated list of kuberentes components to verify and wait for after starting a cluster. defaults to "apiserver,system_pods", available options: "apiserver,system_pods,default_sa" . other acceptable values are 'all' or 'none', 'true' and 'false'
```

### one question this PR does not solve.
do we need to respect the wait flags in a restart ?
the current minikube HEAD does not care about wait=false in a restart.
https://github.com/kubernetes/minikube/blob/5806bb01f26a73233f34a3fbf7473ebf8f8d28d8/pkg/minikube/bootstrapper/kubeadm/kubeadm.go#L497

### CODE REVIEW

File: pkg/minikube/bootstrapper/bsutil/kverify/default_sa.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+		}
+		for _, sa := range sas.Items {
+			if sa.Name == "default" {
+				glog.Infof("found service account: %q", sa.Name)
+				return nil
+			}
+		}
+		return fmt.Errorf("couldn't find default service account")
+	}
+	if err := retry.Expo(saReady, 500*time.Millisecond, 30*time.Second); err != nil {
```
Comment (tstromberg) [inline pkg/minikube/bootstrapper/bsutil/kverify/default_sa.go:50]: 30 seems very short, especially for an explicitly set wait.

File: pkg/minikube/bootstrapper/bsutil/kverify/kverify.go [hunk 2]
```diff
@@ -44,86 +34,31 @@ import (
...
-	})
-	if err != nil {
-		return fmt.Errorf("apiserver process never appeared")
-	}
-	glog.Infof("duration metric: took %s to wait for apiserver process to appear ...", time.Since(start))
-	return nil
-}
+const (
+	// minLogCheckTime how long to wait before spamming error logs to console
+	minLogCheckTime = 30 * time.Second
```
Comment (tstromberg) [inline pkg/minikube/bootstrapper/bsutil/kverify/kverify.go:39]: Leave this at 60.

File: pkg/minikube/bootstrapper/bsutil/kverify/kverify.go [hunk 3]
```diff
@@ -44,86 +34,31 @@ import (
...
-		"kube-controller-manager",
-		"kube-proxy",
-		"kube-scheduler",
-	}
+// DefaultWaitsKeys is list of all default components to wait for
+var DefaultWaitsKeys = []string{APIServerWait, SystemPodsWait}
 
-	found := map[string]bool{}
+// NoWaitsCompos is map of componets to wait for if specified 'none' or 'false'
+var NoWaitsCompos = map[string]bool{APIServerWait: false, SystemPodsWait: false, DefaultServiceAccountWait: false}
```
Comment (tstromberg) [inline pkg/minikube/bootstrapper/bsutil/kverify/kverify.go:55]: Name this "NoWaitComponents".

Also, shouldn't this just be: map[string]bool{}

File: cmd/minikube/cmd/start.go [hunk 4]
```diff
@@ -1194,3 +1196,45 @@ func getKubernetesVersion(old *config.ClusterConfig) string {
...
+
+	waitFlags, err := cmd.Flags().GetStringSlice(waitComponents)
+	if err != nil {
+		glog.Infof("failed to get wait from flags, will use default wait components : %+v", kverify.DefaultWaits)
+		return kverify.DefaultWaits
+	}
+
+	// before minikube 1.9.0, wait flag was boolean
+	if (len(waitFlags) == 1 && waitFlags[0] == "true") || (len(waitFlags) == 1 && waitFlags[0] == "all") {
+		return kverify.AllWaitsCompos
```
[the remark this answers is not in the document]
Comment (medyagh) [reply cmd/minikube/cmd/start.go:1216]: suggestion for better short name components?

File: pkg/minikube/bootstrapper/bsutil/kverify/kverify.go (hunk 3 above)
Comment (medyagh) [reply pkg/minikube/bootstrapper/bsutil/kverify/kverify.go:55]: in the future we might have something super small to verify, even if they say no wait.
I find this more clear and cleaner to be explicit. plus makes the callers of this code look nicer too

File: cmd/minikube/cmd/start.go [hunk 5]
```diff
@@ -1194,3 +1196,45 @@ func getKubernetesVersion(old *config.ClusterConfig) string {
...
+	}
+
+	// respecting legacy flag format --wait=false
+	// before minikube 1.9.0, wait flag was boolean
+	if (len(waitFlags) == 1 && waitFlags[0] == "false") || len(waitFlags) == 1 && waitFlags[0] == "none" {
+		return kverify.NoWaitsCompos
```
[the remark this answers is not in the document]
Comment (medyagh) [reply cmd/minikube/cmd/start.go]: done thanks

File: cmd/minikube/cmd/start.go [hunk 6]
```diff
@@ -1194,3 +1196,45 @@ func getKubernetesVersion(old *config.ClusterConfig) string {
...
+	}
+
+	// before minikube 1.9.0, wait flag was boolean
+	if (len(waitFlags) == 1 && waitFlags[0] == "true") || (len(waitFlags) == 1 && waitFlags[0] == "all") {
+		return kverify.AllWaitsCompos
+	}
+
+	// respecting legacy flag format --wait=false
+	// before minikube 1.9.0, wait flag was boolean
+	if (len(waitFlags) == 1 && waitFlags[0] == "false") || len(waitFlags) == 1 && waitFlags[0] == "none" {
```
[the remark this answers is not in the document]
Comment (medyagh) [reply cmd/minikube/cmd/start.go:1227]: thanks done.

**Pull request merged**